### PR TITLE
MM-47049 : Migrate "components/setting_item_max.jsx" and its associated tests to Typescript

### PR DESCRIPTION
**Pull request by cannalee90**
**Summary**

migrate component/setting_item_max.tsx and its test to typescript

**Ticket Link**

Fixes https://github.com/mattermost/mattermost-server/issues/21140
Fixes https://mattermost.atlassian.net/browse/MM-47049

Related Pull Requeests

NA

Screenshots

NA

Release Note

NA

### CODE REVIEW

File: components/user_settings/advanced/user_settings_advanced.tsx [hunk 1]
```diff
@@ -450,9 +450,9 @@ export default class AdvancedSettingsDisplay extends React.PureComponent<Props,
                         </fieldset>,
                     ]}
                     setting={Preferences.UNREAD_SCROLL_POSITION}
-                    submit={this.handleSubmit}
```
Comment (M-ZubairAhmed) [inline components/user_settings/advanced/user_settings_advanced.tsx:453]: why was this renamed?

File: components/setting_item_max.tsx [hunk 2]
```diff
@@ -135,27 +57,40 @@ export default class SettingItemMax extends React.PureComponent {
         document.removeEventListener('keydown', this.onKeyDown);
     }
 
-    onKeyDown = (e) => {
-        if (this.props.shiftEnter && e.keyCode === Constants.KeyCodes.ENTER && e.shiftKey) {
+    onKeyDown = (e: any) => {
```
Comment (M-ZubairAhmed) [inline components/setting_item_max.tsx:60]: let me more specific than any

**Review by M-ZubairAhmed — CHANGES_REQUESTED** [review]

**Comment by M-ZubairAhmed** [comment]
@cannalee90 can you please resolve conflicts

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 1 above)
Comment (cannalee90) [reply components/user_settings/advanced/user_settings_advanced.tsx:453]: submit handle many kinds of callback which have different signature. I had been working ok until migration. 

 In order to handle this kind of situation, I think it is better to separate two functions one is `submit?: null | ((setting?: string) => void) | (() => Promise<void>);` the other is this `submitWithSettings?: (settings: string[]) => void;`

File: components/setting_item_max.tsx (hunk 2 above)
Comment (cannalee90) [reply components/setting_item_max.tsx:60]: ok I change from any to KeyboardEvent

**Comment by cannalee90** [comment]
> @cannalee90 can you please resolve conflicts

rebase finished

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 1 above)
Comment (M-ZubairAhmed) [reply components/user_settings/advanced/user_settings_advanced.tsx:453]: i see but we need to change the tests to accomodate these changes. i suggest spending a bit more time to see if we can work with just `submit` and if not then let me know we can work something out

Comment (cannalee90) [reply components/user_settings/advanced/user_settings_advanced.tsx:453]: I got it. thanks a lot

File: components/setting_item_max.tsx [hunk 3]
```diff
@@ -1,131 +1,53 @@
...
 
 import SaveButton from 'components/save_button';
 import Constants from 'utils/constants';
 import {isKeyPressed} from 'utils/utils';
 
-export default class SettingItemMax extends React.PureComponent {
+export type Props = {
+    inputs: ReactNode;
+    containerStyle?: string;
+    clientError?: string | React.ReactNode | null;
```
Comment (michelengelen) [inline components/setting_item_max.tsx:14]: `ReactNode` is already typed with `string` and `null` (in addition to some other types), so you can just replace this with a single `React.ReactNode`

File: components/setting_item_max.tsx [hunk 4]
```diff
@@ -1,131 +1,53 @@
...
+    serverError?: {server_error_id: string; message: string} | string | null;
+    extraInfo?: string | JSX.Element | null;
```
Comment (michelengelen) [inline components/setting_item_max.tsx:16]: quite the same goes for `JSX.Element` which also has `string` in the union type already, so you can remove the extra `string` here

File: components/setting_item_max.tsx [hunk 5]
```diff
@@ -1,131 +1,53 @@
...
+    infoPosition?: string;
+    section?: string;
+    updateSection: (section: string) => void;
+    submitWithSettings?: (settings: string[]) => void;
+    submit?: null | ((setting?: string) => void) | (() => Promise<void>);
```
Comment (michelengelen) [inline components/setting_item_max.tsx:21]: I don't think we need `null` here

File: components/setting_item_max.tsx [hunk 6]
```diff
@@ -1,131 +1,53 @@
...
 
     componentDidMount() {
-        if (this.settingList.current) {
-            const focusableElements = this.settingList.current.querySelectorAll('.btn:not(.save-button):not(.btn-cancel), input.form-control, select, textarea, [tabindex]:not([tabindex="-1"])');
-            if (focusableElements.length > 0) {
-                focusableElements[0].focus();
-            } else {
-                this.settingList.current.focus();
-            }
+        const focusableElements = this.settingList?.current?.querySelectorAll?.('.btn:not(.save-button):not(.btn-cancel), input.form-control, select, textarea, [tabindex]:not([tabindex="-1"])');
```
Comment (michelengelen) [inline components/setting_item_max.tsx:48]: the first optional is not needed: 
```suggestion
        const focusableElements = this.settingList.current?.querySelectorAll?.('.btn:not(.save-button):not(.btn-cancel), input.form-control, select, textarea, [tabindex]:not([tabindex="-1"])');
```

File: components/setting_item_max.tsx [hunk 7]
```diff
@@ -1,131 +1,53 @@
...
+        if (focusableElements && focusableElements.length > 0) {
+            (focusableElements[0] as HTMLElement).focus();
         }
```
Comment (michelengelen) [inline components/setting_item_max.tsx:51]:
```suggestion
        if (focusableElements?.length > 0) {
            (focusableElements[0] as HTMLElement).focus();
        }
```

File: components/setting_item_max.tsx [hunk 8]
```diff
@@ -135,27 +57,46 @@ export default class SettingItemMax extends React.PureComponent {
...
             return;
         }
-        if (this.props.disableEnterSubmit !== true && isKeyPressed(e, Constants.KeyCodes.ENTER) && this.props.submit && e.target.tagName !== 'SELECT' && e.target.parentElement && e.target.parentElement.className !== 'react-select__input' && !e.target.classList.contains('btn-cancel') && this.settingList.current && this.settingList.current.contains(e.target)) {
-            this.handleSubmit(e);
+        if (this.props.disableEnterSubmit !== true &&
+            isKeyPressed(e, Constants.KeyCodes.ENTER) &&
+            this.props.submit &&
+            e.target
+        ) {
+            const target = e.target as HTMLElement;
```
Comment (michelengelen) [inline components/setting_item_max.tsx:69]: the casting to `HTMLElement` should not be needed here

File: components/setting_item_max.tsx [hunk 9]
```diff
@@ -135,27 +57,46 @@ export default class SettingItemMax extends React.PureComponent {
...
+            if (target.tagName !== 'SELECT' &&
+                target.parentElement &&
+                target.parentElement.className !== 'react-select__input' &&
+                !target.classList.contains('btn-cancel') &&
+                this.settingList.current &&
+                this.settingList.current.contains(target)
+            ) {
+                this.handleSubmit(e as any);
```
Comment (michelengelen) [inline components/setting_item_max.tsx:77]: please don't use `any` and instead properly type the event (which should not be needed, since it is of type `KeyboardEvent`, which should also be expected by the `handleSubmit` function)

File: components/setting_item_max.tsx [hunk 10]
```diff
@@ -135,27 +57,46 @@ export default class SettingItemMax extends React.PureComponent {
...
+            }
         }
     }
 
-    handleSubmit = (e) => {
+    handleSubmit = (e: React.MouseEvent<HTMLButtonElement>) => {
```
Comment (michelengelen) [inline components/setting_item_max.tsx:82]: please add the `KeyboardEvent` here as well

File: components/setting_item_max.tsx [hunk 11]
```diff
@@ -135,27 +57,46 @@ export default class SettingItemMax extends React.PureComponent {
...
         e.preventDefault();
 
-        if (this.props.setting) {
+        if (!this.props.submit) {
+            return;
+        }
+
+        if (this.props.submitWithSettings) {
+            this.props.submitWithSettings([this.props.setting!]);
+        } else if (this.props.setting) {
```
Comment (michelengelen) [inline components/setting_item_max.tsx:91]: I don't really get why we have this extra submit handler here that accepts an array of settings.
Could you elaborate why that is here?

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 1 above)
Comment (michelengelen) [reply components/user_settings/advanced/user_settings_advanced.tsx:453]: I am the same opinion. We should keep the name and change the function signature instead. we now have two separate props that essentially do the same, but with different parameters. Could you please adjust the signature on all occurences to accept a `string[]` instead?

**Review by michelengelen — CHANGES_REQUESTED** [review]
A few things that need clarification, but all in all looks solid.

File: components/user_settings/advanced/user_settings_advanced.tsx (hunk 1 above)
Comment (cannalee90) [reply components/user_settings/advanced/user_settings_advanced.tsx:453]: ok but I should expand my changes for this task. I'm afraid that my changes are getting bigger and bigger. 

If you don't mind for that I would try

**Comment by M-ZubairAhmed** [comment]
Hey @cannalee90 i very much apologize for not seeing this earlier but it looks like the same ticket has been duplicated. PR for this is already ready for QA https://github.com/mattermost/mattermost-webapp/pull/11329
We are looking into right now if there other duplicates tickets which were opened to avoid wasting valuable time of contributors like you.

**Comment by cannalee90** [comment]
@M-ZubairAhmed It's ok. thank you for your notice
